dapp_runner/descriptor/dapp.py

```diff
@@ -154,24 +180,27 @@ class MetaDescriptor:
     author: str = ""
     version: str = ""
 
-    def __init__(self, **kwargs):
-        for f in fields(self):
-            if f.name in kwargs:
-                setattr(self, f.name, f.type(kwargs.pop(f.name)))
-        if kwargs:
-            logger.debug("Unrecognized `meta` fields: %s", kwargs)
-
 
-@dataclass
-class DappDescriptor(BaseDescriptor["DappDescriptor"]):
+class DappDescriptor(BaseModel):
     """Root dapp descriptor for the Dapp Runner."""
 
     payloads: Dict[str, PayloadDescriptor]
     nodes: Dict[str, ServiceDescriptor]
-    networks: Dict[str, NetworkDescriptor] = field(default_factory=dict)
-    meta: MetaDescriptor = field(default_factory=MetaDescriptor)
+    networks: Dict[str, NetworkDescriptor] = Field(default_factory=dict)
+    meta: MetaDescriptor = Field(default_factory=MetaDescriptor)
 
-    _dependency_graph: networkx.DiGraph = field(init=False)
+    _dependency_graph: networkx.DiGraph = PrivateAttr()
+
+    class Config:  # noqa: D106
+        extra = "forbid"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.__validate_nodes()
+        self.__implicit_proxy_init()
+        self.__implicit_vpn()
+        self.__implicit_manifest_support()
+        self._resolve_dependencies()
```

Comment:
As this works, it could escape few behaviors as [pydantic's author mentioned](https://github.com/pydantic/pydantic/issues/691#issuecomment-515409301). This looks like not so complicated logic that could be implemented in pydantic's per-field validators. Also, back in [my pydantic-like DTOs](https://github.com/golemfactory/yapapi/pull/1103) I've moved default values from parent models down to it's children. In result, I got simpler parent models. Consider similar approach for.

Reply:
as mentioned on a call, these checks are not per-field and cannot be moved to the children